crates/ra_hir/src/ty.rs

```diff
@@ -889,20 +914,59 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             }
             Expr::BinaryOp { lhs, rhs, op } => match op {
                 Some(op) => {
-                    let subtype_expectation = match op {
+                    let lhs_expectation = match op {
                         BinaryOp::BooleanAnd | BinaryOp::BooleanOr => {
                             Expectation::has_type(Ty::Bool)
                         }
                         _ => Expectation::none(),
                     };
-                    let _lhs_ty = self.infer_expr(*lhs, &subtype_expectation)?;
-                    let _rhs_ty = self.infer_expr(*rhs, &subtype_expectation)?;
+                    let lhs_ty = self.infer_expr(*lhs, &lhs_expectation)?;
+                    // TODO: find implementation of trait corresponding to operation
+                    // symbol and resolve associated `Output` type
+                    let rhs_expectation = match op {
```

Comment:
maybe this big match could also be extracted into a function

Reply:
Good idea. Will do.